In [1]:
import pandas as pd
import numpy as np

In [3]:
stores = pd.read_csv("D:\ALabs\Python for Data Science\STUDY MATERIAL\Data Sets\stores.csv")
stores.head()

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   
2    STR103       Super Bazar  Super Market    Delhi           22.8   
3    STR104      Super Market  Super Market    Delhi           21.4   
4    STR105     Central Store  Super Market    Delhi           18.7   

   Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
0         60       160.0              110            3.90       2.620   
1         60       160.0              110            3.90       2.875   
2         40       108.0               93            3.85       2.320   
3         60       258.0              110            3.08       3.215   
4         80       360.0              175            3.15       3.440   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  
0          16.46         0               1       4             4  
1          17.02         0               1       4             4  
2          18.61         1               1       4             1  
3          19.44         1               0       3             1  
4          17.02         0               0       3             2

#### Data Analysis steps:

1. Requirement gathering

2. import packages & dataset---> Dat availability

3. EDA:
        i. Structured : .info()[ .size, .ndim, .shape, .columns, .dtypes ]
        ii. Content based: [ .head(), .tail(), .describe(), .nunique() ] 
        
4. Data cleaning | wrangling| munging | prep:
        i. structured
        ii. content-based
        
5. Data combining:
        join|merge: pd.merge()
        concat: pd.concat(axis = )

### 4. DATA CLEANING | DATA PREP :

### I. STRUCTURED-BASED:

#### a. Subsetting:

* Selecting few variables that are needed for the analysis.
* Subsetting can be done using:

        i. data.<col_name> ----> easiest method
        ii. data['col_name'] : works with DI
        iii. .iloc[ : ] : works with DI
        iv. .loc[ : ] : works with UDI

In [4]:
stores.head()

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   
2    STR103       Super Bazar  Super Market    Delhi           22.8   
3    STR104      Super Market  Super Market    Delhi           21.4   
4    STR105     Central Store  Super Market    Delhi           18.7   

   Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
0         60       160.0              110            3.90       2.620   
1         60       160.0              110            3.90       2.875   
2         40       108.0               93            3.85       2.320   
3         60       258.0              110            3.08       3.215   
4         80       360.0              175            3.15       3.440   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  
0          16.46         0               1       4             4  
1          17.02         0               1       4             4  
2          18.61         1               1       4             1  
3          19.44         1               0       3             1  
4          17.02         0               0       3             2

#### 1. Get the first variable from the dataset

In [14]:
# 1 method:
stores.StoreCode.head()

# 2nd method:
stores['StoreCode'].head()

# 3rd method: using .loc[]
stores.loc[ : , 'StoreCode'].head()

# 4th method: using .iloc[]
stores.iloc[:, 0].head()

0    STR101
1    STR102
2    STR103
3    STR104
4    STR105
Name: StoreCode, dtype: object

#### 2. Get the Location data from the stores dataset using .iloc[] & .loc[]

In [19]:
# using .iloc[]
stores.iloc[ : ,  3]

# using .loc[]
stores.loc[ : ,  'Location'].head(3)

0    Delhi
1    Delhi
2    Delhi
Name: Location, dtype: object

#### 3. Get the data from the first 4 columns of the stores dataset

In [27]:
# 1st method:
stores[['StoreCode', 'StoreName', 'StoreType', 'Location']].head()

#2nd method: using .loc[]
stores.loc[ :, ['StoreCode', 'StoreName', 'StoreType', 'Location'] ]

# 3rd method: using .iloc[]
stores.iloc[ :, 0:4 ]

StoreCode         StoreName     StoreType Location
0     STR101  Electronics Zone  Electronincs    Delhi
1     STR102      Apparel Zone       Apparel    Delhi
2     STR103       Super Bazar  Super Market    Delhi
3     STR104      Super Market  Super Market    Delhi
4     STR105     Central Store  Super Market    Delhi
5     STR106      Apparel Zone       Apparel    Delhi
6     STR107     Fashion Bazar       Apparel    Delhi
7     STR108     Digital Bazar  Electronincs    Delhi
8     STR109  Electronics Zone  Electronincs  Chennai
9     STR110      Apparel Zone       Apparel  Chennai
10    STR111       Super Bazar  Super Market  Chennai
11    STR112      Super Market  Super Market  Chennai
12    STR113     Central Store  Super Market  Chennai
13    STR114      Apparel Zone       Apparel  Chennai
14    STR115     Fashion Bazar       Apparel  Chennai
15    STR116     Digital Bazar  Electronincs  Chennai
16    STR117  Electronics Zone  Electronincs   Mumbai
17    STR118      Apparel Zone       Apparel   Mumbai
18    STR119       Super Bazar  Super Market   Mumbai
19    STR120      Super Market  Super Market   Mumbai
20    STR121     Central Store  Super Market   Mumbai
21    STR122      Apparel Zone       Apparel   Mumbai
22    STR123     Fashion Bazar       Apparel   Mumbai
23    STR124     Digital Bazar  Electronincs   Mumbai
24    STR125  Electronics Zone  Electronincs  Kolkata
25    STR126      Apparel Zone       Apparel  Kolkata
26    STR127       Super Bazar  Super Market  Kolkata
27    STR128      Super Market  Super Market  Kolkata
28    STR129     Central Store  Super Market  Kolkata
29    STR130      Apparel Zone       Apparel  Kolkata
30    STR131     Fashion Bazar       Apparel  Kolkata
31    STR132     Digital Bazar  Electronincs  Kolkata

##### In the above question, where multiple variables need to be extracted, always use [ ]

#### b. Calculated Fields: 

* **To get a new variable in the dataset, there are different ways:**
    
        i. data['new_col_name'] = value/operation
        
        ii. using .loc:
                data.loc[ :, 'new_col_name'] = value/operation
                
        iii. .insert()
        
        iv. .assign()
    
**.iloc[ ] can't be used as the control is not with the user**

#### 1. Get the data that finds the total Acquistion cost for each customer

In [31]:
stores.head()

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   
2    STR103       Super Bazar  Super Market    Delhi           22.8   
3    STR104      Super Market  Super Market    Delhi           21.4   
4    STR105     Central Store  Super Market    Delhi           18.7   

   Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
0         60       160.0              110            3.90       2.620   
1         60       160.0              110            3.90       2.875   
2         40       108.0               93            3.85       2.320   
3         60       258.0              110            3.08       3.215   
4         80       360.0              175            3.15       3.440   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  
0          16.46         0               1       4             4  
1          17.02         0               1       4             4  
2          18.61         1               1       4             1  
3          19.44         1               0       3             1  
4          17.02         0               0       3             2

In [35]:
# 1st method:

stores['TotalAcqCost'] = stores.Total_Customers * stores.AcqCostPercust
stores.head(3)

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   
2    STR103       Super Bazar  Super Market    Delhi           22.8   

   Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
0         60       160.0              110            3.90       2.620   
1         60       160.0              110            3.90       2.875   
2         40       108.0               93            3.85       2.320   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0          16.46         0               1       4             4        429.00  
1          17.02         0               1       4             4        429.00  
2          18.61         1               1       4             1        358.05

In [146]:
# 2nd method: using .loc[]
stores.loc[: ,'TotalAcqCost1'] = stores.Total_Customers * stores.AcqCostPercust
stores.head(3)

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   

   TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  BasketSize  \
0       160.0              110         429.00            3.90       2.620   
1       160.0              110         429.00            3.90       2.875   
2       108.0               93         358.05            3.85       2.320   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0          16.46         0               1       4             4   
1          17.02         0               1       4             4   
2          18.61         1               1       4             1   

   TotalAcqCost  TotalAcqCost1  
0        429.00         429.00  
1        429.00         429.00  
2        358.05         358.05

In [147]:
# 3rd method: using .insert()

stores.insert(8, 'TotalAcqCost2', stores.Total_Customers * stores.AcqCostPercust )

ValueError: cannot insert TotalAcqCost2, already exists

In [148]:
stores.head(3)

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   

   TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  BasketSize  \
0       160.0              110         429.00            3.90       2.620   
1       160.0              110         429.00            3.90       2.875   
2       108.0               93         358.05            3.85       2.320   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0          16.46         0               1       4             4   
1          17.02         0               1       4             4   
2          18.61         1               1       4             1   

   TotalAcqCost  TotalAcqCost1  
0        429.00         429.00  
1        429.00         429.00  
2        358.05         358.05

In [43]:
# 4th method: using .assign()

stores.assign(col1 = stores.Total_Customers * stores.AcqCostPercust, col2 = stores.Total_Customers * stores.AcqCostPercust )

StoreCode         StoreName     StoreType Location  OperatingCost  \
0     STR101  Electronics Zone  Electronincs    Delhi           21.0   
1     STR102      Apparel Zone       Apparel    Delhi           21.0   
2     STR103       Super Bazar  Super Market    Delhi           22.8   
3     STR104      Super Market  Super Market    Delhi           21.4   
4     STR105     Central Store  Super Market    Delhi           18.7   
5     STR106      Apparel Zone       Apparel    Delhi           18.1   
6     STR107     Fashion Bazar       Apparel    Delhi           14.3   
7     STR108     Digital Bazar  Electronincs    Delhi           24.4   
8     STR109  Electronics Zone  Electronincs  Chennai           22.8   
9     STR110      Apparel Zone       Apparel  Chennai           19.2   
10    STR111       Super Bazar  Super Market  Chennai           17.8   
11    STR112      Super Market  Super Market  Chennai           16.4   
12    STR113     Central Store  Super Market  Chennai           17.3   
13    STR114      Apparel Zone       Apparel  Chennai           15.2   
14    STR115     Fashion Bazar       Apparel  Chennai           10.4   
15    STR116     Digital Bazar  Electronincs  Chennai           10.4   
16    STR117  Electronics Zone  Electronincs   Mumbai           14.7   
17    STR118      Apparel Zone       Apparel   Mumbai           32.4   
18    STR119       Super Bazar  Super Market   Mumbai           30.4   
19    STR120      Super Market  Super Market   Mumbai           33.9   
20    STR121     Central Store  Super Market   Mumbai           21.5   
21    STR122      Apparel Zone       Apparel   Mumbai           15.5   
22    STR123     Fashion Bazar       Apparel   Mumbai           15.2   
23    STR124     Digital Bazar  Electronincs   Mumbai           13.3   
24    STR125  Electronics Zone  Electronincs  Kolkata           19.2   
25    STR126      Apparel Zone       Apparel  Kolkata           27.3   
26    STR127       Super Bazar  Super Market  Kolkata           26.0   
27    STR128      Super Market  Super Market  Kolkata           30.4   
28    STR129     Central Store  Super Market  Kolkata           15.8   
29    STR130      Apparel Zone       Apparel  Kolkata           19.7   
30    STR131     Fashion Bazar       Apparel  Kolkata           15.0   
31    STR132     Digital Bazar  Electronincs  Kolkata           21.4   

    Staff_Cnt  TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  \
0          60       160.0              110         429.00            3.90   
1          60       160.0              110         429.00            3.90   
2          40       108.0               93         358.05            3.85   
3          60       258.0              110         338.80            3.08   
4          80       360.0              175         551.25            3.15   
5          60       225.0              105         289.80            2.76   
6          80       360.0              245         786.45            3.21   
7          40       146.7               62         228.78            3.69   
8          40       140.8               95         372.40            3.92   
9          60       167.6              123         482.16            3.92   
10         60       167.6              123         482.16            3.92   
11         80       275.8              180            NaN             NaN   
12         80       275.8              180            NaN             NaN   
13         80       275.8              180            NaN             NaN   
14         80       472.0              205         600.65            2.93   
15         80       460.0              215         645.00            3.00   
16         80       440.0              230         742.90            3.23   
17         40        78.7               66         269.28            4.08   
18         40        75.7               52         256.36            4.93   
19         40        71.1               65         274.30            4.22   
20       

In [44]:
stores.head(3)

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   
2    STR103       Super Bazar  Super Market    Delhi           22.8   

   Staff_Cnt  TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  \
0         60       160.0              110         429.00            3.90   
1         60       160.0              110         429.00            3.90   
2         40       108.0               93         358.05            3.85   

   BasketSize  ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0       2.620          16.46         0               1       4             4   
1       2.875          17.02         0               1       4             4   
2       2.320          18.61         1               1       4             1   

   TotalAcqCost  TotalAcqCost1  
0        429.00         429.00  
1        429.00         429.00  
2        358.05         358.05

#### c.Renaming Variables:

* Done using the function: 
        .rename()

In [45]:
stores.columns

Index(['StoreCode', 'StoreName', 'StoreType', 'Location', 'OperatingCost',
       'Staff_Cnt', 'TotalSales', 'Total_Customers', 'TotalAcqCost2',
       'AcqCostPercust', 'BasketSize', 'ProfitPercust', 'OwnStore',
       'OnlinePresence', 'Tenure', 'StoreSegment', 'TotalAcqCost',
       'TotalAcqCost1'],
      dtype='object')

In [47]:
stores.head(2)

StoreCode         StoreName     StoreType Location  OperatingCost  \
0    STR101  Electronics Zone  Electronincs    Delhi           21.0   
1    STR102      Apparel Zone       Apparel    Delhi           21.0   

   Staff_Cnt  TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  \
0         60       160.0              110          429.0             3.9   
1         60       160.0              110          429.0             3.9   

   BasketSize  ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0       2.620          16.46         0               1       4             4   
1       2.875          17.02         0               1       4             4   

   TotalAcqCost  TotalAcqCost1  
0         429.0          429.0  
1         429.0          429.0

In [51]:
stores.rename(columns = {'Location' : 'Region'}, inplace = True)
stores.head()

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   
3    STR104      Super Market  Super Market  Delhi           21.4         60   
4    STR105     Central Store  Super Market  Delhi           18.7         80   

   TotalSales  Total_Customers  TotalAcqCost2  AcqCostPercust  BasketSize  \
0       160.0              110         429.00            3.90       2.620   
1       160.0              110         429.00            3.90       2.875   
2       108.0               93         358.05            3.85       2.320   
3       258.0              110         338.80            3.08       3.215   
4       360.0              175         551.25            3.15       3.440   

   ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0          16.46         0               1       4             4   
1          17.02         0               1       4             4   
2          18.61         1               1       4             1   
3          19.44         1               0       3             1   
4          17.02         0               0       3             2   

   TotalAcqCost  TotalAcqCost1  
0        429.00         429.00  
1        429.00         429.00  
2        358.05         358.05  
3        338.80         338.80  
4        551.25         551.25

#### d. Removal of the Variables:

* For deleting any variable, function used is .drop()

In [149]:
stores.drop(columns = ['TotalAcqCost2', 'TotalAcqCost1'], inplace = True)

In [150]:
stores.head(3)

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
0       160.0              110            3.90       2.620          16.46   
1       160.0              110            3.90       2.875          17.02   
2       108.0               93            3.85       2.320          18.61   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0         0               1       4             4        429.00  
1         0               1       4             4        429.00  
2         1               1       4             1        358.05

### II. CONTENT BASED DATA CLEANING :

#### a. Filters:

* Types of filters:
        i. manual
        ii. conditional


##### Manual Filtering:

In [61]:
# get the first 4 records from the dataset:

print( stores.head(4) )  # MANUAL FILTER
print( stores.tail(4) )  # MANUAL FILTER


  StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   
3    STR104      Super Market  Super Market  Delhi           21.4         60   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
0       160.0              110            3.90       2.620          16.46   
1       160.0              110            3.90       2.875          17.02   
2       108.0               93            3.85       2.320          18.61   
3       258.0              110            3.08       3.215          19.44   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0         0               1       4             4        429.00  
1         0               1       4             4        429.00  
2         1    

In [63]:
# using .iloc[] , .loc[]:
stores.iloc[ 0: 3, :]

stores.loc[ 0 : 2, :]

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
0       160.0              110            3.90       2.620          16.46   
1       160.0              110            3.90       2.875          17.02   
2       108.0               93            3.85       2.320          18.61   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0         0               1       4             4        429.00  
1         0               1       4             4        429.00  
2         1               1       4             1        358.05

#### 1. Get the records from the stores dataset where the row numbers are 2, 5, 10, 1, 20

In [67]:
stores.iloc[[2, 5, 10, 1, 20]]

stores.loc[[2, 5, 10, 1, 20] ]

StoreCode      StoreName     StoreType   Region  OperatingCost  Staff_Cnt  \
2     STR103    Super Bazar  Super Market    Delhi           22.8         40   
5     STR106   Apparel Zone       Apparel    Delhi           18.1         60   
10    STR111    Super Bazar  Super Market  Chennai           17.8         60   
1     STR102   Apparel Zone       Apparel    Delhi           21.0         60   
20    STR121  Central Store  Super Market   Mumbai           21.5         40   

    TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
2        108.0               93            3.85       2.320          18.61   
5        225.0              105            2.76       3.460          20.22   
10       167.6              123            3.92       3.440          18.90   
1        160.0              110            3.90       2.875          17.02   
20       120.1               97            3.70       2.465          20.01   

    OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
2          1               1       4             1        358.05  
5          1               0       3             1        289.80  
10         1               0       4             4        482.16  
1          0               1       4             4        429.00  
20         1               0       3             1        358.90

#### 2. Get the random 10 records from the stores dataset

In [85]:
stores.shape[0]

32

In [89]:
stores.iloc[ np.random.randint(0, stores.shape[0], 10) ] 

StoreCode         StoreName     StoreType   Region  OperatingCost  \
25    STR126      Apparel Zone       Apparel  Kolkata           27.3   
1     STR102      Apparel Zone       Apparel    Delhi           21.0   
31    STR132     Digital Bazar  Electronincs  Kolkata           21.4   
0     STR101  Electronics Zone  Electronincs    Delhi           21.0   
23    STR124     Digital Bazar  Electronincs   Mumbai           13.3   
12    STR113     Central Store  Super Market  Chennai           17.3   
19    STR120      Super Market  Super Market   Mumbai           33.9   
28    STR129     Central Store  Super Market  Kolkata           15.8   
11    STR112      Super Market  Super Market  Chennai           16.4   
31    STR132     Digital Bazar  Electronincs  Kolkata           21.4   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
25         40        79.0               66            4.08       1.935   
1          60       160.0              110            3.90       2.875   
31         40       121.0              109            4.11       2.780   
0          60       160.0              110            3.90       2.620   
23         80       350.0              245            3.73       3.840   
12         80       275.8              180             NaN       3.730   
19         40        71.1               65            4.22       1.835   
28         80       351.0              264            4.22       3.170   
11         80       275.8              180             NaN       4.070   
31         40       121.0              109            4.11       2.780   

    ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
25          18.90         1               1       4             1   
1           17.02         0               1       4             4   
31          18.60         1               1       4             2   
0           16.46         0               1       4             4   
23          15.41         0               0       3             4   
12          17.60         0               0       3             3   
19          19.90         1               1       4             1   
28          14.50         0               1       5             4   
11          17.40         0               0       3             3   
31          18.60         1               1       4             2   

    TotalAcqCost  
25        269.28  
1         429.00  
31        447.99  
0         429.00  
23        913.85  
12           NaN  
19        274.30  
28       1114.08  
11           NaN  
31        447.99

##### Conditional Filtering:

* It can be done using:

        i. .loc[]
        ii. .iloc[]
        iii. data[<col_name]

#### 1. Get all the records from the dataset, where the Location is Delhi.

In [98]:
# 1st method
stores[stores['Region'] == 'Delhi']

#2nd method: using .iloc[]
stores[ stores.loc[ : , 'Region' ] == 'Delhi' ] 

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   
3    STR104      Super Market  Super Market  Delhi           21.4         60   
4    STR105     Central Store  Super Market  Delhi           18.7         80   
5    STR106      Apparel Zone       Apparel  Delhi           18.1         60   
6    STR107     Fashion Bazar       Apparel  Delhi           14.3         80   
7    STR108     Digital Bazar  Electronincs  Delhi           24.4         40   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
0       160.0              110            3.90       2.620          16.46   
1       160.0              110            3.90       2.875          17.02   
2       108.0               93            3.85       2.320          18.61   
3       258.0              110            3.08       3.215          19.44   
4       360.0              175            3.15       3.440          17.02   
5       225.0              105            2.76       3.460          20.22   
6       360.0              245            3.21       3.570          15.84   
7       146.7               62            3.69       3.190          20.00   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0         0               1       4             4        429.00  
1         0               1       4             4        429.00  
2         1               1       4             1        358.05  
3         1               0       3             1        338.80  
4         0               0       3             2        551.25  
5         1               0       3             1        289.80  
6         0               0       3             4        786.45  
7         1               0       4             2        228.78

#### 2. Get the observations with total Sales > overall average sales

In [100]:
stores.TotalSales.mean()

230.72187500000004

In [103]:
stores[ stores.TotalSales >= stores.TotalSales.mean() ]

StoreCode         StoreName     StoreType   Region  OperatingCost  \
3     STR104      Super Market  Super Market    Delhi           21.4   
4     STR105     Central Store  Super Market    Delhi           18.7   
6     STR107     Fashion Bazar       Apparel    Delhi           14.3   
11    STR112      Super Market  Super Market  Chennai           16.4   
12    STR113     Central Store  Super Market  Chennai           17.3   
13    STR114      Apparel Zone       Apparel  Chennai           15.2   
14    STR115     Fashion Bazar       Apparel  Chennai           10.4   
15    STR116     Digital Bazar  Electronincs  Chennai           10.4   
16    STR117  Electronics Zone  Electronincs   Mumbai           14.7   
21    STR122      Apparel Zone       Apparel   Mumbai           15.5   
22    STR123     Fashion Bazar       Apparel   Mumbai           15.2   
23    STR124     Digital Bazar  Electronincs   Mumbai           13.3   
24    STR125  Electronics Zone  Electronincs  Kolkata           19.2   
28    STR129     Central Store  Super Market  Kolkata           15.8   
30    STR131     Fashion Bazar       Apparel  Kolkata           15.0   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
3          60       258.0              110            3.08       3.215   
4          80       360.0              175            3.15       3.440   
6          80       360.0              245            3.21       3.570   
11         80       275.8              180             NaN       4.070   
12         80       275.8              180             NaN       3.730   
13         80       275.8              180             NaN       3.780   
14         80       472.0              205            2.93       5.250   
15         80       460.0              215            3.00       5.424   
16         80       440.0              230            3.23       5.345   
21         80       318.0              150            2.76       3.520   
22         80       304.0              150            3.15       3.435   
23         80       350.0              245            3.73       3.840   
24         80       400.0              175            3.08       3.845   
28         80       351.0              264            4.22       3.170   
30         80       301.0              335            3.54       3.570   

    ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
3           19.44         1               0       3             1   
4           17.02         0               0       3             2   
6           15.84         0               0       3             4   
11          17.40         0               0       3             3   
12          17.60         0               0       3             3   
13          18.00         0               0       3             3   
14          17.98         0               0       3             4   
15          17.82         0               0       3             4   
16          17.42         0               0       3             4   
21          16.87         0               0       3             2   
22          17.30         0               0       3             2   
23          15.41         0               0       3             4   
24          17.05         0               0       3             2   
28          14.50         0               1       5             4   
30          14.60         0               1       5             4   

    TotalAcqCost  
3         338.80  
4         551.25  
6         786.45  
11           NaN  
12           NaN  
13           NaN  
14        600.65  
15        645.00  
16        742.90  
21        414.00  
22        472.50  
23        913.85  
24        539.00  
28       1114.08  
30       1185.90

#### 3. Get the records where the location is Delhi & the total sales > overall average sales

In [105]:
stores[ (stores.Region == 'Delhi') & 
       (stores.TotalSales >= stores.TotalSales.mean()) ]

StoreCode      StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
3    STR104   Super Market  Super Market  Delhi           21.4         60   
4    STR105  Central Store  Super Market  Delhi           18.7         80   
6    STR107  Fashion Bazar       Apparel  Delhi           14.3         80   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
3       258.0              110            3.08       3.215          19.44   
4       360.0              175            3.15       3.440          17.02   
6       360.0              245            3.21       3.570          15.84   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
3         1               0       3             1        338.80  
4         0               0       3             2        551.25  
6         0               0       3             4        786.45

#### 4. Get the records where the location is Delhi & the total sales > overall avg sales. Print the Store code, Store name, Location & Total sales

In [109]:
stores.loc[ (stores.Region == 'Delhi') & 
       (stores.TotalSales >= stores.TotalSales.mean()), ['StoreCode', 'StoreName', 'Region', 'TotalSales'] ]

StoreCode      StoreName Region  TotalSales
3    STR104   Super Market  Delhi       258.0
4    STR105  Central Store  Delhi       360.0
6    STR107  Fashion Bazar  Delhi       360.0

#### 5. Get the records where the location is Delhi and Mumbai.
                              OR
#### Get the records where the location is Delhi or Mumbai.

In [112]:
stores[ (stores.Region == 'Delhi') | (stores.Region == 'Mumbai') ]

StoreCode         StoreName     StoreType  Region  OperatingCost  \
0     STR101  Electronics Zone  Electronincs   Delhi           21.0   
1     STR102      Apparel Zone       Apparel   Delhi           21.0   
2     STR103       Super Bazar  Super Market   Delhi           22.8   
3     STR104      Super Market  Super Market   Delhi           21.4   
4     STR105     Central Store  Super Market   Delhi           18.7   
5     STR106      Apparel Zone       Apparel   Delhi           18.1   
6     STR107     Fashion Bazar       Apparel   Delhi           14.3   
7     STR108     Digital Bazar  Electronincs   Delhi           24.4   
16    STR117  Electronics Zone  Electronincs  Mumbai           14.7   
17    STR118      Apparel Zone       Apparel  Mumbai           32.4   
18    STR119       Super Bazar  Super Market  Mumbai           30.4   
19    STR120      Super Market  Super Market  Mumbai           33.9   
20    STR121     Central Store  Super Market  Mumbai           21.5   
21    STR122      Apparel Zone       Apparel  Mumbai           15.5   
22    STR123     Fashion Bazar       Apparel  Mumbai           15.2   
23    STR124     Digital Bazar  Electronincs  Mumbai           13.3   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
0          60       160.0              110            3.90       2.620   
1          60       160.0              110            3.90       2.875   
2          40       108.0               93            3.85       2.320   
3          60       258.0              110            3.08       3.215   
4          80       360.0              175            3.15       3.440   
5          60       225.0              105            2.76       3.460   
6          80       360.0              245            3.21       3.570   
7          40       146.7               62            3.69       3.190   
16         80       440.0              230            3.23       5.345   
17         40        78.7               66            4.08       2.200   
18         40        75.7               52            4.93       1.615   
19         40        71.1               65            4.22       1.835   
20         40       120.1               97            3.70       2.465   
21         80       318.0              150            2.76       3.520   
22         80       304.0              150            3.15       3.435   
23         80       350.0              245            3.73       3.840   

    ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
0           16.46         0               1       4             4   
1           17.02         0               1       4             4   
2           18.61         1               1       4             1   
3           19.44         1               0       3             1   
4           17.02         0               0       3             2   
5           20.22         1               0       3             1   
6           15.84         0               0       3             4   
7           20.00         1               0       4             2   
16          17.42         0               0       3             4   
17          19.47         1               1       4             1   
18          18.52         1               1       4             2   
19          19.90         1               1       4             1   
20          20.01         1               0       3             1   
21          16.87         0               0       3             2   
22          17.30         0               0       3             2   
23          15.41         0               0       3             4   

    TotalAcqCost  
0         429.00  
1         429.00  
2         358.05  
3         338.80  
4         551.25  
5         289.80  
6         786.45  
7         228.78  
16        742.90  
17        269.28  
18        256.36  
19        274.30  
20        358.90  
21        414.00  
22        472.50  
23        913.85

#### c. Sorting:

* Data can be arranged either in ASC or DESC order.
* Easy to analyse the data/ vizualization improves.

* 2 ways to do the sorting:
        i. .sort_index()
        ii. .sort_values()

#### 1. Sort the data as per the row index numbers.

In [118]:
stores.sort_index(axis = 0).head()

StoreCode         StoreName     StoreType Region  OperatingCost  Staff_Cnt  \
0    STR101  Electronics Zone  Electronincs  Delhi           21.0         60   
1    STR102      Apparel Zone       Apparel  Delhi           21.0         60   
2    STR103       Super Bazar  Super Market  Delhi           22.8         40   
3    STR104      Super Market  Super Market  Delhi           21.4         60   
4    STR105     Central Store  Super Market  Delhi           18.7         80   

   TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
0       160.0              110            3.90       2.620          16.46   
1       160.0              110            3.90       2.875          17.02   
2       108.0               93            3.85       2.320          18.61   
3       258.0              110            3.08       3.215          19.44   
4       360.0              175            3.15       3.440          17.02   

   OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
0         0               1       4             4        429.00  
1         0               1       4             4        429.00  
2         1               1       4             1        358.05  
3         1               0       3             1        338.80  
4         0               0       3             2        551.25

#### 2. Sort the data as per the TotalSales in Ascending order

In [120]:
stores.sort_values(by = 'TotalSales').head()

StoreCode     StoreName     StoreType   Region  OperatingCost  Staff_Cnt  \
19    STR120  Super Market  Super Market   Mumbai           33.9         40   
18    STR119   Super Bazar  Super Market   Mumbai           30.4         40   
17    STR118  Apparel Zone       Apparel   Mumbai           32.4         40   
25    STR126  Apparel Zone       Apparel  Kolkata           27.3         40   
27    STR128  Super Market  Super Market  Kolkata           30.4         40   

    TotalSales  Total_Customers  AcqCostPercust  BasketSize  ProfitPercust  \
19        71.1               65            4.22       1.835          19.90   
18        75.7               52            4.93       1.615          18.52   
17        78.7               66            4.08       2.200          19.47   
25        79.0               66            4.08       1.935          18.90   
27        95.1              113            3.77       1.513          16.90   

    OwnStore  OnlinePresence  Tenure  StoreSegment  TotalAcqCost  
19         1               1       4             1        274.30  
18         1               1       4             2        256.36  
17         1               1       4             1        269.28  
25         1               1       4             1        269.28  
27         1               1       5             2        426.01

#### 2. Sort the data as per the TotalSales in Descending order

In [124]:
stores.sort_values(by = 'TotalSales', ascending =  False).head(4)

StoreCode         StoreName     StoreType   Region  OperatingCost  \
14    STR115     Fashion Bazar       Apparel  Chennai           10.4   
15    STR116     Digital Bazar  Electronincs  Chennai           10.4   
16    STR117  Electronics Zone  Electronincs   Mumbai           14.7   
24    STR125  Electronics Zone  Electronincs  Kolkata           19.2   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
14         80       472.0              205            2.93       5.250   
15         80       460.0              215            3.00       5.424   
16         80       440.0              230            3.23       5.345   
24         80       400.0              175            3.08       3.845   

    ProfitPercust  OwnStore  OnlinePresence  Tenure  StoreSegment  \
14          17.98         0               0       3             4   
15          17.82         0               0       3             4   
16          17.42         0               0       3             4   
24          17.05         0               0       3             2   

    TotalAcqCost  
14        600.65  
15        645.00  
16        742.90  
24        539.00

#### 3. Sort the dataset by Location & TotalSales in Descending order

In [126]:
stores.sort_values(by = ['Region', 'TotalSales'], ascending =  False)

StoreCode         StoreName     StoreType   Region  OperatingCost  \
16    STR117  Electronics Zone  Electronincs   Mumbai           14.7   
23    STR124     Digital Bazar  Electronincs   Mumbai           13.3   
21    STR122      Apparel Zone       Apparel   Mumbai           15.5   
22    STR123     Fashion Bazar       Apparel   Mumbai           15.2   
20    STR121     Central Store  Super Market   Mumbai           21.5   
17    STR118      Apparel Zone       Apparel   Mumbai           32.4   
18    STR119       Super Bazar  Super Market   Mumbai           30.4   
19    STR120      Super Market  Super Market   Mumbai           33.9   
24    STR125  Electronics Zone  Electronincs  Kolkata           19.2   
28    STR129     Central Store  Super Market  Kolkata           15.8   
30    STR131     Fashion Bazar       Apparel  Kolkata           15.0   
29    STR130      Apparel Zone       Apparel  Kolkata           19.7   
31    STR132     Digital Bazar  Electronincs  Kolkata           21.4   
26    STR127       Super Bazar  Super Market  Kolkata           26.0   
27    STR128      Super Market  Super Market  Kolkata           30.4   
25    STR126      Apparel Zone       Apparel  Kolkata           27.3   
4     STR105     Central Store  Super Market    Delhi           18.7   
6     STR107     Fashion Bazar       Apparel    Delhi           14.3   
3     STR104      Super Market  Super Market    Delhi           21.4   
5     STR106      Apparel Zone       Apparel    Delhi           18.1   
0     STR101  Electronics Zone  Electronincs    Delhi           21.0   
1     STR102      Apparel Zone       Apparel    Delhi           21.0   
7     STR108     Digital Bazar  Electronincs    Delhi           24.4   
2     STR103       Super Bazar  Super Market    Delhi           22.8   
14    STR115     Fashion Bazar       Apparel  Chennai           10.4   
15    STR116     Digital Bazar  Electronincs  Chennai           10.4   
11    STR112      Super Market  Super Market  Chennai           16.4   
12    STR113     Central Store  Super Market  Chennai           17.3   
13    STR114      Apparel Zone       Apparel  Chennai           15.2   
9     STR110      Apparel Zone       Apparel  Chennai           19.2   
10    STR111       Super Bazar  Super Market  Chennai           17.8   
8     STR109  Electronics Zone  Electronincs  Chennai           22.8   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
16         80       440.0              230            3.23       5.345   
23         80       350.0              245            3.73       3.840   
21         80       318.0              150            2.76       3.520   
22         80       304.0              150            3.15       3.435   
20         40       120.1               97            3.70       2.465   
17         40        78.7               66            4.08       2.200   
18         40        75.7               52            4.93       1.615   
19         40        71.1               65            4.22       1.835   
24         80       400.0              175            3.08       3.845   
28         80       351.0              264            4.22       3.170   
30         80       301.0              335            3.54       3.570   
29         60       145.0              175            3.62       2.770   
31         40       121.0              109            4.11       2.780   
26         40       120.3               91            4.43       2.140   
27         40        95.1              113            3.77       1.513   
25         40        79.0               66            4.08       1.935   
4          80       360.0              175            3.15       3.440   
6          80       360.0              245            3.21       3.570   
3          60       258.0              110            3.08       3.215   
5          60       225.0              105            2.76       3.460   
0          60       160.0              110            3.90       2.620  

#### 4. Sort the dataset by Location in Ascending order & TotalSales in Descending order

In [128]:
stores.sort_values( by = ['Region', 'TotalSales'], ascending = [True, False])

StoreCode         StoreName     StoreType   Region  OperatingCost  \
14    STR115     Fashion Bazar       Apparel  Chennai           10.4   
15    STR116     Digital Bazar  Electronincs  Chennai           10.4   
11    STR112      Super Market  Super Market  Chennai           16.4   
12    STR113     Central Store  Super Market  Chennai           17.3   
13    STR114      Apparel Zone       Apparel  Chennai           15.2   
9     STR110      Apparel Zone       Apparel  Chennai           19.2   
10    STR111       Super Bazar  Super Market  Chennai           17.8   
8     STR109  Electronics Zone  Electronincs  Chennai           22.8   
4     STR105     Central Store  Super Market    Delhi           18.7   
6     STR107     Fashion Bazar       Apparel    Delhi           14.3   
3     STR104      Super Market  Super Market    Delhi           21.4   
5     STR106      Apparel Zone       Apparel    Delhi           18.1   
0     STR101  Electronics Zone  Electronincs    Delhi           21.0   
1     STR102      Apparel Zone       Apparel    Delhi           21.0   
7     STR108     Digital Bazar  Electronincs    Delhi           24.4   
2     STR103       Super Bazar  Super Market    Delhi           22.8   
24    STR125  Electronics Zone  Electronincs  Kolkata           19.2   
28    STR129     Central Store  Super Market  Kolkata           15.8   
30    STR131     Fashion Bazar       Apparel  Kolkata           15.0   
29    STR130      Apparel Zone       Apparel  Kolkata           19.7   
31    STR132     Digital Bazar  Electronincs  Kolkata           21.4   
26    STR127       Super Bazar  Super Market  Kolkata           26.0   
27    STR128      Super Market  Super Market  Kolkata           30.4   
25    STR126      Apparel Zone       Apparel  Kolkata           27.3   
16    STR117  Electronics Zone  Electronincs   Mumbai           14.7   
23    STR124     Digital Bazar  Electronincs   Mumbai           13.3   
21    STR122      Apparel Zone       Apparel   Mumbai           15.5   
22    STR123     Fashion Bazar       Apparel   Mumbai           15.2   
20    STR121     Central Store  Super Market   Mumbai           21.5   
17    STR118      Apparel Zone       Apparel   Mumbai           32.4   
18    STR119       Super Bazar  Super Market   Mumbai           30.4   
19    STR120      Super Market  Super Market   Mumbai           33.9   

    Staff_Cnt  TotalSales  Total_Customers  AcqCostPercust  BasketSize  \
14         80       472.0              205            2.93       5.250   
15         80       460.0              215            3.00       5.424   
11         80       275.8              180             NaN       4.070   
12         80       275.8              180             NaN       3.730   
13         80       275.8              180             NaN       3.780   
9          60       167.6              123            3.92       3.440   
10         60       167.6              123            3.92       3.440   
8          40       140.8               95            3.92       3.150   
4          80       360.0              175            3.15       3.440   
6          80       360.0              245            3.21       3.570   
3          60       258.0              110            3.08       3.215   
5          60       225.0              105            2.76       3.460   
0          60       160.0              110            3.90       2.620   
1          60       160.0              110            3.90       2.875   
7          40       146.7               62            3.69       3.190   
2          40       108.0               93            3.85       2.320   
24         80       400.0              175            3.08       3.845   
28         80       351.0              264            4.22       3.170   
30         80       301.0              335            3.54       3.570   
29         60       145.0              175            3.62       2.770   
31         40       121.0              109            4.11       2.780  

#### 5. Get the data of locations Delhi & Kolkata. Display the output of store name, location & operating cost in Ascending order of operating cost

In [131]:
stores.loc[ (stores.Region == 'Delhi') |( stores.Region == 'Kolkata'), ['StoreName', 'Region', 'OperatingCost'] ].sort_values(by = 'OperatingCost')

StoreName   Region  OperatingCost
6      Fashion Bazar    Delhi           14.3
30     Fashion Bazar  Kolkata           15.0
28     Central Store  Kolkata           15.8
5       Apparel Zone    Delhi           18.1
4      Central Store    Delhi           18.7
24  Electronics Zone  Kolkata           19.2
29      Apparel Zone  Kolkata           19.7
0   Electronics Zone    Delhi           21.0
1       Apparel Zone    Delhi           21.0
3       Super Market    Delhi           21.4
31     Digital Bazar  Kolkata           21.4
2        Super Bazar    Delhi           22.8
7      Digital Bazar    Delhi           24.4
26       Super Bazar  Kolkata           26.0
25      Apparel Zone  Kolkata           27.3
27      Super Market  Kolkata           30.4

###  Handling Duplicates:
***part of Content based Data Cleaning***

1. Detection:
                .duplicated()
2. Treatment:
              .drop_duplicates()
           

In [133]:
score = pd.read_csv("D:\ALabs\Python for Data Science\STUDY MATERIAL\Data Sets\Score.csv")
score

Student  Section  Test1  Test2  Final
0   Capalleti        1     94     91     87
1      Dubose        2     51     65     91
2      Engles        1     95     97     97
3       Grant        2     63     75     80
4     Krupski        2     80     76     71
5   Lundsford        1     92     40     86
6      Mcbane        1     75     78     72
7   Capalleti        1     94     65     87
8      Dubose        2     51     65     91
9      Engles        1     95     97     97
10      Grant        2     63     75     80
11    Krupski        2     80     76     71
12  Lundsford        1     92     40     86
13     Mcbane        1     75     78     72

##### Detection:
        .duplicated()

In [134]:
score.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8      True
9      True
10     True
11     True
12     True
13     True
dtype: bool

In [138]:
score.duplicated(keep = False)

0     False
1      True
2      True
3      True
4      True
5      True
6      True
7     False
8      True
9      True
10     True
11     True
12     True
13     True
dtype: bool

##### Treatment: drop_duplicstes()

In [139]:
score.drop_duplicates()

Student  Section  Test1  Test2  Final
0  Capalleti        1     94     91     87
1     Dubose        2     51     65     91
2     Engles        1     95     97     97
3      Grant        2     63     75     80
4    Krupski        2     80     76     71
5  Lundsford        1     92     40     86
6     Mcbane        1     75     78     72
7  Capalleti        1     94     65     87

In [144]:
score.drop_duplicates(subset = 'Student', keep = 'first', inplace = True)
score

Student  Section  Test1  Test2  Final
0  Capalleti        1     94     91     87
1     Dubose        2     51     65     91
2     Engles        1     95     97     97
3      Grant        2     63     75     80
4    Krupski        2     80     76     71
5  Lundsford        1     92     40     86
6     Mcbane        1     75     78     72

###  Handling Outliers:

**Outlier:**
        
        The records which lie outside the acceptable range, where the cutoff can be calculated using any of the detection methods.
        
* Detection:

    Detection can be using the statistical functions:
            i. %tile  method: commonly  used, its reasy to work with, no calculation involved
            ii. IQR method: lot of calculations are required
            iii. STD. method

* Treatment:
        
        Remove the outliers using pd.clip(lower = lc, upper =  uc)

#### %tile Method:

        Select the 1% of the records from both the ends as an outlier.
        lc = p1
        uc = p99
        
#### IQR method: Inter Quartile Range

        IQR =  P75 - P25  or Q3 - Q1
        
        p0(min), p100(max), p50, p25, p75 -----> .quantile(q = 0.5, 0.25, 0.75)
        
        lc =  p25 - 1.5 * IQR
        uc =  p75 + 1.5 * IQR
        
#### STD. method:

        lc =  mean - 2.5 * std
        uc =  mean + 2.5 * std